### ISSUE TITLE
UAI '20 | Sliced score matching: A scalable approach to density and score estimation.

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#32]
Song Y, Garg S, Shi J, et al. [Sliced score matching: A scalable approach to density and score estimation](http://proceedings.mlr.press/v115/song20a/song20a.pdf).

**Comment by NorbertZheng** [comment]
## Related Reference
 - Hyvärinen A, Dayan P. [Estimation of non-normalized statistical models by score matching](https://www.jmlr.org/papers/volume6/hyvarinen05a/hyvarinen05a.pdf).
 - Martens J, Sutskever I, Swersky K. [Estimating the hessian by back-propagating curvature](https://arxiv.org/pdf/1206.6464.pdf).
 - Kingma D P, Welling M. [Auto-encoding variational bayes](https://arxiv.org/pdf/1312.6114.pdf).
 - Tolstikhin I, Bousquet O, Gelly S, et al. [Wasserstein auto-encoders](https://arxiv.org/pdf/1711.01558.pdf).
 - Song Y, Ermon S. [Generative modeling by estimating gradients of the data distribution](https://proceedings.neurips.cc/paper/2019/file/3001ef257407d5a371a96dcd947c7d93-Paper.pdf).

**Comment by NorbertZheng** [comment]
## Overview
An overview for our UAI 2019 paper on Sliced Score Matching. We show
 - how to use random projection to scale up score matching - a classic method to learn unnormalized probabilistic models - to high-dimensional data.

Theoretically, sliced score matching produces a consistent and asymptotic normal estimator under some regularity conditions. We apply sliced score matching to training deep energy-based models, learning VAEs with implicit encoders, and training Wasserstein Auto-Encoders (WAEs).

**Comment by NorbertZheng** [comment]
## Unnormalized probability models
Let $f_{\theta}(x)$ be a real-valued function on $x \in \mathbb{R}^{D}$. We can define a probability model from $f_{\theta}(x)$, using the following formula

$$
p_{\theta}(x)=\frac{e^{-f_{\theta}(x)}}{Z_{\theta}}.
$$

Here we assume that

$$
Z_{\theta}=\int e^{-f_{\theta}(x)}dx
$$

exists and call it the partition function. Typically, it is intractable to evaluate the partition function, and $Z_{\theta}$ is an unknown quantity. The probability model $p_{\theta}(x)$ is called an unnormalized probability model, because the normalizing constant $Z_{\theta}$ is unknown.

**Comment by NorbertZheng** [comment]
## Learning unnormalized models with score matching
The de facto standard for learning a probability model is the maximum likelihood (MLE). However, it is not suitable for learning unnormalized probability models, because the likelihood function

$$
\log p_\theta(\mathbf{x}) = -f_\theta(\mathbf{x}) - \log Z_\theta
$$

depends on the intractable partition function $Z_{\theta}$.

[Score matching](https://www.jmlr.org/papers/volume6/hyvarinen05a/hyvarinen05a.pdf) bypasses the intractable $Z_{\theta}$ by only considering the scores, which is defined as
 - the gradient of the log-density w.r.t. the random variable, e.g. data, instead of parameter.

For example, the score of $p_{\theta}(x)$ is

$$
\nabla_{x}logp_{\theta}(x)=-\nabla_{x}f_{\theta}(x).
$$

Note that the score of an unnormalized probability model $\nabla_{x}logp_{\theta}(x)$ does not depend on the intractable partition function $Z_{\theta}$. Instead of using likelihood as the objective, score matching is based on the following objective

$$
\frac{1}{2}\mathbb{E_{p_{data}}}\left[\vert\vert\nabla_{x}logp_{data}(x)-\nabla_{x}logp_{\theta}(x)\vert\vert_{2}^{2}\right]
$$

**Comment by NorbertZheng** [comment]
## Sliced score matching
So far, we know that score matching can be used to learn unnormalized models. We are particularly interested in learning deep energy-based models, a special kind of unnormalized models where $f_{\theta}(x)$ is parameterized by a deep neural network. In order to use score matching for learning deep energy-based models, we have to compute

$$
\| \nabla_\mathbf{x} \log p_\theta(\mathbf{x})\|_2^2 = \| \nabla_\mathbf{x} f_\theta(\mathbf{x})\|^2_2
$$

, and

$$
\operatorname{tr}( \nabla_{\mathbf{x}}^2 \log p_\theta(\mathbf{x})) = -\operatorname{tr}( \nabla_{\mathbf{x}}^2 f_\theta(\mathbf{x}))
$$

The former can be computed by one simple backpropagation of $f_{\theta}(x)$. The latter, however, requires much more backpropagations to compute. As studied in [previous work](https://arxiv.org/pdf/1206.6464.pdf), computing $\operatorname{tr}( \nabla_{\mathbf{x}}^2 f_\theta(\mathbf{x}))$ requires a number of backpropagation that is proportional to the data dimension $D$. Therefore, score matching is not scalable when learning deep energy-based models on high-dimensional data.

**Comment by NorbertZheng** [comment]
We propose sliced score matching to greatly scale up the computation of score matching. The motivating idea is that one-dimensional data distribution is much easier to estimate for score matching. We propose to
 - project the scores onto random directions, such that the vector fields of scores of the data and model distribution become scalar fields, e.g. force all vectors onto one dimension?

We then compare the scalar fields to determine how far the model distribution is from the data distribution. It is clear to see that the two vector fields are equivalent if and only if their scalar fields corresponding to projections onto all directions are the same. So we can project them onto many directions, e.g. sampleing?

Mathematically, we denote $v$ as the random projection disrection, and $p_{v}$ as its distribution. The random projected version of Fisher divergence is

$$
\frac{1}{2}\mathbb{E}\_{p_\text{data}}[(\mathbf{v}^\intercal \nabla_\mathbf{x} \log p_\text{data}(\mathbf{x}) - \mathbf{v}^\intercal \nabla_\mathbf{x} \log p_\theta(\mathbf{x}) )^2],
$$

, which we name as the sliced Fisher divergence. Unfortunately, sliced Fisher divergence has the same problem as Fisher divergence, due to the unknown data score function $\nabla_{x}logp_{data}(x)$. We therefore play the same trick of integration by parts, as done in score matching, to obtain the following tractable alternative form

$$
\mathbb{E}\_{p_\text{data}}\bigg[\mathbf{v}^\intercal \nabla_{\mathbf{x}}^2 \log p_\theta(\mathbf{x})\mathbf{v} + \frac{1}{2} (\mathbf{v}^\intercal\nabla_\mathbf{x} \log p_\theta(\mathbf{x}))^2 \bigg] + \text{const},
$$

, which is our sliced score matching objective. Again, $\mathbf{v}^\intercal\nabla_\mathbf{x} \log p_\theta(\mathbf{x})$ can be computed by one backpropagation for deep energy-based models. The first item of the objective again involves Hessian, but it is in the form of Hessian-vector products, which can be computed with $\mathcal{O}(1)$ backpropagations. Therefore, the computation of sliced score matching does not depend on the dimension of data, and is much more scalable for training deep energy-based models on high dimensional datasets.

**Comment by NorbertZheng** [comment]
## Theoretical guarantees of learning with sliced score matching
Suppose our data points $\{ \mathbf{x}_1, \mathbf{x}_2, \cdots, \mathbf{x}_N\}$ are i.i.d. samples from the data distribution $p_{data}$. For each data point $\mathbf{x}_i$, we randomly draw $M$ random projection directions $\{ \mathbf{v}_{i1}, \mathbf{v}_{i2}, \cdots, \mathbf{v}_{iM}\} \sim p_\mathbf{v}$. The sliced score matching objective can be estimated with empirical averages, giving rise to the following finite-sample estimator:

$$
\frac{1}{NM} \sum_{i=1}^N \sum_{j=1}^M \bigg[\mathbf{v}_{ij}^\intercal \nabla_{\mathbf{x}}^2 \log p_\theta(\mathbf{x}_i)\mathbf{v}_{ij} + \frac{1}{2} (\mathbf{v}_{ij}^\intercal\nabla_\mathbf{x} \log p_\theta(\mathbf{x}_i))^2 \bigg]
$$

**Comment by NorbertZheng** [comment]
## Experiments
Sliced score matching has numerous applications. For example,
 - Density estimation. It can be used to scalably train deep energy-based models on high-dimensional data.
 - Score estimation. The sliced matching objective can be used to estimate the score of any distribution from which samples can be efficiently obtained, which we call score estimation. It can be used to learn [Variational Auto-Encoders (VAEs)](https://arxiv.org/pdf/1312.6114.pdf) where the encoder is an implicit distribution. It can also be used to learn [Wasserstein Auto-Encoders (WAEs)](https://arxiv.org/pdf/1711.01558.pdf).
 - Score-based generative modeling. It can be used to estimate scores of the data distribution and produce samples directly with Langevin dynamics.
